substrate/frame/session/src/lib.rs

```diff
@@ -836,10 +860,39 @@ impl<T: Config> Pallet<T> {
 			.ok_or(Error::<T>::NoAssociatedValidatorId)?;
 
 		ensure!(frame_system::Pallet::<T>::can_inc_consumer(account), Error::<T>::NoAccount);
-		let old_keys = Self::inner_set_keys(&who, keys)?;
-		if old_keys.is_none() {
+		
+		// Load keys once to avoid redundant storage reads
+		let old_keys = Self::load_keys(&who);
+		let is_new_registration = old_keys.is_none();
+		
+		// For new registrations, ensure the account has enough funds for the deposit
+		if is_new_registration {
+			let deposit = T::KeyDeposit::get();
+			let reason = T::HoldReason::get();
+			ensure!(
```

Comment:
If this is called from within a transaction, which AFAIK it is, no need to do this check first -- you can directly call `T::Currency::hold(&reason, account, deposit)?;`. If it fails, everything is rolled back. 

Recall that FRAME is transactional ;)

Reply:
Thanks :D